benchmark/Benchmark.cpp

```diff
@@ -0,0 +1,125 @@
+}
+
+// ____________________________________________________________________________
+BenchmarkRegister::BenchmarkRegister(const std::vector<std::function<void(BenchmarkRecords*)>>& benchmarks) {
+  for (const auto& benchmark: benchmarks) {
+    _registerdBenchmarks.push_back(benchmark);
+  }
+}
+
+// ____________________________________________________________________________
+float BenchmarkRecords::measureTimeOfFunction(
+    const std::function<void()>& functionToMeasure) const {
+  ad_utility::Timer benchmarkTimer;
+     
+  benchmarkTimer.start();
+  functionToMeasure();
+  benchmarkTimer.stop();
+
+  return benchmarkTimer.secs();
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addSingleMeasurment(const std::string descriptor,
+      const std::function<void()>& functionToMeasure) {
+    _singleMeasurments.push_back(RecordEntry{descriptor,
+        measureTimeOfFunction(functionToMeasure)});
+}
+
+// ____________________________________________________________________________
+const std::vector<BenchmarkRecords::RecordEntry>&
+   BenchmarkRecords::getSingleMeasurments() const{
+  return _singleMeasurments;
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addGroup(const std::string descriptor) {
+  // Is there already a group with this descriptor? If so, that is not allowed.
+  AD_CHECK(_recordGroups.find(descriptor) == _recordGroups.end());
+
+  // There is no group, so create one without any entries and add them to
+  // the hash map.
+  _recordGroups[descriptor] = BenchmarkRecords::RecordGroup{descriptor, {}};
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addToExistingGroup(const std::string& groupDescriptor,
+    const std::string descriptor,
+    const std::function<void()>& functionToMeasure) {
+  // Does the group exist?
+  auto groupEntry = _recordGroups.find(groupDescriptor);
+  AD_CHECK(groupEntry != _recordGroups.end());
+
+  // Add the descriptor and measured time to the group.
+  groupEntry->second.entries.push_back(
+      RecordEntry{descriptor, measureTimeOfFunction(functionToMeasure)});
+}
+
+// ____________________________________________________________________________
+const ad_utility::HashMap<std::string, BenchmarkRecords::RecordGroup>&
+   BenchmarkRecords::getGroups() const {
+  return _recordGroups;
+}
+
+// ____________________________________________________________________________
+void BenchmarkRecords::addTable(const std::string descriptor,
+    const std::vector<std::string> rowNames,
+    const std::vector<std::string> columnNames) {
+  // Is there already a table with this descriptor? If so, that is not allowed.
+  AD_CHECK(_recordTables.find(descriptor) == _recordTables.end());
+
+  // Add a new entry.
+  _recordTables[descriptor] = BenchmarkRecords::RecordTable{descriptor,
+    rowNames, columnNames, {}};
```

Comment:
Make the `rowNames` and `columnNames` not const, but just `vector<..>`, and then `std::move()` them here (you have to save their sizes though if you still need them below).

Reply:
I already changed them to `const vector<...>&`, so that is no longer necessary.